src/Chainweb/Pact/PactService.hs

```diff
@@ -622,21 +625,62 @@ execNewGenesisBlock miner newTrans = withDiscardedBatch $
 execLocal
     :: CanReadablePayloadCas tbl
     => ChainwebTransaction
-    -> PactServiceM tbl (P.CommandResult P.Hash)
-execLocal cmd = withDiscardedBatch $ do
+    -> Bool
+      -- ^ preflight flag
+    -> Maybe Word64
+      -- ^ rewind depth
+    -> PactServiceM tbl (Either MetadataValidationFailure (P.CommandResult P.Hash))
+execLocal cwtx preflight rdepth = withDiscardedBatch $ do
     PactServiceEnv{..} <- ask
+
+    let !cmd = payloadObj <$> cwtx
+        !pm = publicMetaOf cmd
+
     mc <- getInitCache
-    pd <- getTxContext (publicMetaOf $! payloadObj <$> cmd)
+    ctx <- getTxContext pm
     spv <- use psSpvSupport
+
+    let rewindHeight
+          | Just d <- rdepth = Just $ BlockHeight d
+          -- when no height is defined, treat
+          -- withCheckpointerRewind as withCurrentCheckpointer
+          -- (i.e. setting rewind to 0).
+          | otherwise = Just 0
+        rewindHeader = Just $ _tcParentHeader ctx
+
     let execConfig = P.mkExecutionConfig $
             [ P.FlagAllowReadInLocal | _psAllowReadsInLocal ] ++
-            enablePactEvents' pd ++
-            enforceKeysetFormats' pd
+            enablePactEvents' ctx ++
+            enforceKeysetFormats' ctx
         logger = P.newLogger _psLoggers "execLocal"
-    withCurrentCheckpointer "execLocal" $ \(PactDbEnv' pdbenv) -> do
-        r <- liftIO $
-          applyLocal logger _psGasLogger pdbenv chainweb213GasModel pd spv cmd mc execConfig
-        return $! Discard (toHashCommandResult r)
+        initialGas = initialGasOf $ P._cmdPayload cwtx
+
+    withCheckpointerRewind rewindHeight rewindHeader "execLocal" $
+      \(PactDbEnv' pdbenv) -> do
+        --
+        -- if the ?preflight query parameter is set to True, we run the `applyCmd` workflow
+        -- otherwise, we prefer the old (default) behavior. When no preflight flag is
+        -- specified, we run the old behavior. When it is set to true, we also do metadata
+        -- validations.
+        --
+        r <- if preflight
+          then do
+            assertLocalMetadata cmd ctx >>= \case
+              Right{} -> do
+                T2 cr _mc' <- liftIO $ applyCmd
+                  _psVersion logger _psGasLogger pdbenv
+                  noMiner chainweb213GasModel ctx spv cmd
+                  initialGas mc ApplyLocal
+                pure $ Right cr
+              Left e -> pure $ Left e
+          else liftIO $ do
+            cr <- applyLocal
+              logger _psGasLogger pdbenv
+              chainweb213GasModel ctx spv
+              cwtx mc execConfig
+            pure $ Right cr
+
+        return $! Discard (toHashCommandResult <$> r)
```

Comment:
This will have no effect. The construction is already in WHNF form: We know it's the `Discard` constructor!

Reply:
This was as it was before - i can remove the bang if we care to